### ISSUE TITLE
some JUnit tests fail after updating to 1.2.1

### ISSUE THREAD

**Issue by DenBond7** [vanitasvitae/pgpainless#282]
Hi

We have one PR https://github.com/FlowCrypt/flowcrypt-android/pull/1770 with the following changes

![image](https://user-images.githubusercontent.com/2863246/166637639-cc807d85-880d-4970-8f13-f0772f3bb57a.png)

After updating to 1.2.1 some of our tests fail. 

I'm talking about this one [PgpDecryptAndOrVerifyTest](https://github.com/FlowCrypt/flowcrypt-android/blob/a08da60848b6ff829a5a86740b3786d35505f71d/FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpDecryptAndOrVerifyTest.kt)

![image](https://user-images.githubusercontent.com/2863246/166636277-a5c26e79-7ea3-4348-9f72-41991fdad145.png)

Please let me add details about one of them [PgpDecryptAndOrVerifyTest.testDecryptBinaryFileSuccess()](https://github.com/FlowCrypt/flowcrypt-android/blob/a08da60848b6ff829a5a86740b3786d35505f71d/FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpDecryptAndOrVerifyTest.kt#L56)

```kotlin
private fun testDecryptFileSuccess(shouldSrcBeArmored: Boolean) {
    val srcFile = temporaryFolder.createFileWithRandomData(fileSizeInBytes = FileUtils.ONE_MB)
    val sourceBytes = srcFile.readBytes()
    val outputStreamForEncryptedSource = ByteArrayOutputStream()
    PgpEncryptAndOrSign.encryptAndOrSign(
      srcInputStream = ByteArrayInputStream(sourceBytes),
      destOutputStream = outputStreamForEncryptedSource,
      pgpPublicKeyRingCollection = PGPPublicKeyRingCollection(
        listOf(
          KeyRingUtils.publicKeyRingFrom(senderPGPSecretKeyRing),
          KeyRingUtils.publicKeyRingFrom(receiverPGPSecretKeyRing)
        )
      ),
      doArmor = shouldSrcBeArmored// false in our case
    )

    val encryptedBytes = outputStreamForEncryptedSource.toByteArray()
    val outputStreamWithDecryptedData = ByteArrayOutputStream()
    PgpDecryptAndOrVerify.decrypt(
      srcInputStream = ByteArrayInputStream(encryptedBytes),
      destOutputStream = outputStreamWithDecryptedData,
      secretKeys = PGPSecretKeyRingCollection(listOf(receiverPGPSecretKeyRing)),
      protector = allPredefinedKeysProtector
    )

    val decryptedBytesForSender = outputStreamWithDecryptedData.toByteArray()
    Assert.assertArrayEquals(sourceBytes, decryptedBytesForSender)
  }
```

```kotlin
fun TemporaryFolder.createFileWithRandomData(
  fileSizeInBytes: Long,
  fileName: String = UUID.randomUUID().toString(),
  bufferSize: Int = (FileUtils.ONE_KB * 8).toInt()
): File {
  return newFile(fileName).apply {
    RandomAccessFile(this, "rw").apply {
      setLength(fileSizeInBytes)
      outputStream().use { outStream ->
        var overriddenBytesCount = 0L
        val random = Random()
        val buffer = ByteArray(bufferSize)
        while (overriddenBytesCount != fileSizeInBytes) {
          if (fileSizeInBytes - overriddenBytesCount <= bufferSize) {
            val lastSegmentBufferSize = (fileSizeInBytes - overriddenBytesCount).toInt()
            val lastSegmentBuffer = ByteArray(lastSegmentBufferSize)
            random.nextBytes(lastSegmentBuffer)
            outStream.write(lastSegmentBuffer)
            overriddenBytesCount += lastSegmentBufferSize
          } else {
            random.nextBytes(buffer)
            outStream.write(buffer)
            overriddenBytesCount += bufferSize
          }
        }
        outStream.flush()
      }
    }
  }
}
```

```kotlin
@Throws(IOException::class)
  fun encryptAndOrSign(
    srcInputStream: InputStream, destOutputStream: OutputStream,
    pgpPublicKeyRingCollection: PGPPublicKeyRingCollection,
    pgpSecretKeyRingCollection: PGPSecretKeyRingCollection? = null,
    secretKeyRingProtector: SecretKeyRingProtector? = null,
    doArmor: Boolean = false,
    passphrase: Passphrase? = null
  ) {
    srcInputStream.use { srcStream ->
      genEncryptionStream(
        destOutputStream = destOutputStream,
        pgpPublicKeyRingCollection = pgpPublicKeyRingCollection,
        pgpSecretKeyRingCollection = pgpSecretKeyRingCollection,
        secretKeyRingProtector = secretKeyRingProtector,
        doArmor = doArmor,
        passphrase = passphrase
      ).use { encryptionStream ->
        srcStream.copyTo(encryptionStream)
      }
    }
  }
```

```kotlin
  fun genEncryptionStream(
    destOutputStream: OutputStream,
    pgpPublicKeyRingCollection: PGPPublicKeyRingCollection,
    pgpSecretKeyRingCollection: PGPSecretKeyRingCollection?,
    secretKeyRingProtector: SecretKeyRingProtector?,
    doArmor: Boolean,
    passphrase: Passphrase? = null
  ): EncryptionStream {
    val encOpt = EncryptionOptions().apply {
      passphrase?.let { addPassphrase(passphrase) }
      pgpPublicKeyRingCollection.forEach {
        addRecipient(it)
      }
    }

    val producerOptions: ProducerOptions =
      if (passphrase == null && pgpSecretKeyRingCollection?.any() == true) {
        ProducerOptions.signAndEncrypt(encOpt, SigningOptions().apply {
          pgpSecretKeyRingCollection.forEach {
            addInlineSignature(
              secretKeyRingProtector, it, DocumentSignatureType.BINARY_DOCUMENT
            )
          }
        })
      } else {
        ProducerOptions.encrypt(encOpt)
      }

    producerOptions.isAsciiArmor = doArmor

    return PGPainless.encryptAndOrSign()
      .onOutputStream(destOutputStream)
      .withOptions(producerOptions)
  }
```

```kotlin
fun decrypt(
    srcInputStream: InputStream,
    destOutputStream: OutputStream,
    secretKeys: PGPSecretKeyRingCollection,
    protector: SecretKeyRingProtector
  ): OpenPgpMetadata {
    srcInputStream.use { srcStream ->
      destOutputStream.use { outStream ->
        try {
          val decryptionStream = genDecryptionStream(
            srcInputStream = srcStream,
            secretKeys = secretKeys,
            protector = protector
          )
          decryptionStream.use { it.copyTo(outStream) }
          return decryptionStream.result
        } catch (e: Exception) {
          throw processDecryptionException(e)
        }
      }
    }
  }
```

```kotlin
fun genDecryptionStream(
    srcInputStream: InputStream,
    publicKeys: PGPPublicKeyRingCollection? = null,
    secretKeys: PGPSecretKeyRingCollection,
    protector: SecretKeyRingProtector,
    ignoreMdcErrors: Boolean = false
  ): DecryptionStream {
    return PGPainless.decryptAndOrVerify()
      .onInputStream(srcInputStream)
      .withOptions(
        ConsumerOptions()
          .addDecryptionKeys(secretKeys, protector)
          .setMissingKeyPassphraseStrategy(MissingKeyPassphraseStrategy.THROW_EXCEPTION)
          .setIgnoreMDCErrors(ignoreMdcErrors)
          .apply {
            publicKeys?.let { addVerificationCerts(it) }
          }
      )
  }
```

![image](https://user-images.githubusercontent.com/2863246/166636668-fe44b9af-a214-47ed-8fd3-88b33761ad26.png)

It seems decryption for binary source was affected in 1.2.1.

What is strange - this test passes if temporaryFolder.createFileWithRandomData(fileSizeInBytes <~ 800)(around 800 bytes)

**Comment by vanitasvitae** [comment]
Are the tests failing consistently? Or only every now and then? Also, did they succeed previously consistently?

My intention with the changes in 1.2.1 were to improve on this situation, so it's too bad to see that it got worse :sweat_smile:

**Comment by DenBond7** [comment]
> Are the tests failing consistently?

Yes, they are. The tests fail every time.

>Also, did they succeed previously consistently?

 Before 1.2.1 all worked well.

I can identify a commit that may be a reason. Please let me know.

**Comment by IvanPizhenko** [comment]
@vanitasvitae I'm getting following error in another project after updating to PGPainless 1.2.1:
```
'org.bouncycastle.asn1.ASN1BitString org.bouncycastle.asn1.x509.SubjectPublicKeyInfo.getPublicKeyData()'
java.lang.NoSuchMethodError: 'org.bouncycastle.asn1.ASN1BitString org.bouncycastle.asn1.x509.SubjectPublicKeyInfo.getPublicKeyData()'
	at org.bouncycastle.openpgp.operator.jcajce.JcaPGPKeyConverter.getPublicBCPGKey(Unknown Source)
	at org.bouncycastle.openpgp.operator.jcajce.JcaPGPKeyConverter.getPGPPublicKey(Unknown Source)
	at org.bouncycastle.openpgp.operator.jcajce.JcaPGPKeyConverter.getPGPPublicKey(Unknown Source)
	at org.bouncycastle.openpgp.operator.jcajce.JcaPGPKeyPair.getPublicKey(Unknown Source)
	at org.bouncycastle.openpgp.operator.jcajce.JcaPGPKeyPair.<init>(Unknown Source)
	at org.pgpainless.implementation.BcImplementationFactory.jceToBcKeyPair(BcImplementationFactory.java:166)
	at org.pgpainless.implementation.BcImplementationFactory.getPGPKeyPair(BcImplementationFactory.java:134)
	at org.pgpainless.key.generation.KeyRingBuilder.generateKeyPair(KeyRingBuilder.java:312)
	at org.pgpainless.key.generation.KeyRingBuilder.build(KeyRingBuilder.java:156)
	at com.flowcrypt.shared.crypto.Pgp.generateKey(Pgp.kt:311)
	at com.flowcrypt.shared.crypto.PgpTest.set expiration of key that does not expire - for now missing preferred hash algos(PgpTest.kt:285)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

**Comment by vanitasvitae** [comment]
Could you open another ticket for this issue and add more info on how to recreate the problem?

**Comment by DenBond7** [comment]
> I can identify a commit that may be a reason. Please let me know.

I'm adding it just for history

![image](https://user-images.githubusercontent.com/2863246/166887087-c6e60313-de70-4c66-90cf-6c79e5f7052b.png)

![image](https://user-images.githubusercontent.com/2863246/166887134-69713574-aed6-4074-853d-e5711f22d270.png)

![image](https://user-images.githubusercontent.com/2863246/166887165-3ef88694-2784-4884-9ad1-15095adac10b.png)

**Comment by vanitasvitae** [comment]
I just released version 1.2.2 which contains the fixes :)
